Jetpack Boost: Image Guide compatibility improvements

pyronaur opened the pull request:
Introduce portals to break the Popup out of its containers as various plugins affect the Popup differently.

#### Changes proposed in this Pull Request:

* Add portals
* Use a fixed position for the guide

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

#### Jetpack product discussion
pc9hqz-1sC-p2#comment-1104

#### Does this pull request change what data or activity we track or use?
No

### Todo

- [x] Add external link icon pc9hqz-1sC-p2#comment-1095
- [x] Remove the hideous hack in `initialize.ts`
- [x] Address any other feedback in the thread: pc9hqz-1sC-p2

#### Testing instructions:

Install the following plugins and populate them with content:
* Various Gutenberg galleries
* Gutenberg full-width container with a fixed position background image:

This is a typical page that I use for most testing (apart from the plugins listed below):

https://user-images.githubusercontent.com/988095/208727490-6c5c352a-d7a8-47f9-b59c-1839170002fd.mp4




* Jetpack (Jetpack mosaic/tiled gallery, Jetpack image comparison)
* Smart Slider
* Visual Portfolio with Masonry layout:
<img width="1318" alt="CleanShot 2022-12-20 at 19 16 48@2x" src="https://user-images.githubusercontent.com/988095/208726759-231e7cec-2109-41c1-9909-08946df1f96c.png">

pyronaur committed:
Use a fixed position for the guide

pyronaur committed:
Increase padding

pyronaur committed:
Add a Portal component that teleports slot to body

pyronaur committed:
Portal the pop-up to body to fix z-index issues

pyronaur committed:
FIxed a compatibility bug with Jetpack Mosaic gallery

Wrapping caused the object-fit to break, this seems to be a more lightweight approach.

pyronaur committed:
A hacky compatibility fix across plugins

pyronaur committed:
Simplify keeping the Pop-up open using a class

pyronaur committed:
Rename classes

pyronaur commented on projects/plugins/boost/app/features/image-guide/src/initialize.ts:
```diff
@@ -31,8 +30,22 @@ function getClosestContainingAncestor( node: HTMLElement ): HTMLElement | null {
+			elementsWithoutZIndex = [ current ];
 		} else {
-			elementsWithoutZIndex = [];
+			// The idea here was to reset elementsWithoutZIndex array
+			// whenever we encounter an element that has a z-index set.
+			// But that breaks the portfolio in the portfolio plugin layout.
+			// By not resetting the array, we're keeping
+			// tracks of first static element in the tree.
+			// This seems to work out ok for now,
+			// But at this point, the logic here now
+			// has changed from "reset the array when we encounter a z-indexed element"
+			// to "reset when we encounter a non-static element that's absolutely positioned"
+			// elementsWithoutZIndex = [];
```
`@TODO`

dilirity replied:
This has been done in d1f6125 and 9103e82.

thingalon reviewed (dismissed):
I'm ok with these changes, but I think more experimentation is needed. I'm still getting issues in tiled image galleries, and I'm just not sure about the z-index ancestor search logic. I might try to fiddle more with it this week :)

pyronaur committed:
Don't overlay other interactive elements as much as possible

By taking up as little space as possible.

pyronaur committed:
Be a little more clear on what's going on, WIP

Commit:
[not verified] Merge

Commit:
changelog

Commit:
Undo inexplicable change from merge

thingalon reviewed (dismissed):
😮 this is working, and it's working _well_. It works nicely on Tiled Galleries, and I am all for merging this asap.

Just 🌈 glorious!

I merged from trunk and got some weird monorepo ... glitching. Happy to help untangle that, but I've also raised it with crew here: p1673589594309319-slack-CBG1CP4EN

dilirity committed:
Update info block position in tooltip

dilirity commented on docs/testing/regression-checklist/mapping.json:
```diff
@@ -30,7 +30,7 @@
 	},
 	{
-		"regex": "^projects\/plugins\/jetpack\/extensions\/(blocks\/(subscriptions|premium-content)\/|shared\/memberships\.js)",
+		"regex": "^projects/plugins/jetpack/extensions/(blocks/(subscriptions|premium-content)/|shared/memberships.js)",
```
The version in trunk is different.
https://github.com/Automattic/jetpack/blob/21c9c9d586c08e25af92f5de3d8122234f4fb4ff/docs/testing/regression-checklist/mapping.json#L33

dilirity replied:
@pyronaur which version is correct? The other regexes in the file aren't escaped.

pyronaur replied:
I've got no idea what this is and how it got here.

pyronaur replied:
We probably need to revert this

pyronaur replied:
Ah, I see what you mean, yeah - the others aren't escaped, so i think this'll be fine?

dilirity replied:
Prettier formatting did this. Closing.

dilirity committed:
Fix image guide inside links

dilirity committed:
Refactor getClosestContainingAncestor to not use arrays to keep track of current element

dilirity committed:
Refactor getClosestContainingAncestor to improve code readability

dilirity committed:
Disable image guide in iframes

This is done to ensure editors using iframes don't initialize
the image guide in the WP-Admin.

pyronaur committed:
Lints

dilirity committed:
Revert regex update that happened during trunk merge

dilirity reviewed (approved):
Let's merge this.

dereksmart reviewed (approved):
Approving on behalf of the required review for the docs/ update
